api/src/main/java/org/pmiops/workbench/db/dao/DataSetServiceImpl.java

```diff
@@ -171,163 +205,309 @@ public DataSet saveDataSet(
+    } else if (!includesAllParticipants) {
+      queryBuilder
+          .append(" \nWHERE ")
+          .append(personIdQualified)
+          .append(" IN (")
+          .append(cohortQueries)
+          .append(")");
+    }
+
+    final String completeQuery = queryBuilder.toString();
+
+    return buildQueryJobConfiguration(cohortParameters, completeQuery);
+  }
+
+  private void validateConceptSetSelection(Domain domain, List<ConceptSet> conceptSetsSelected) {
+    if (supportsConceptSets(domain)
+        && !conceptSetSelectionIsNonemptyAndEachDomainHasAtLeastOneConcept(conceptSetsSelected)) {
+      throw new BadRequestException("Concept Sets must contain at least one concept");
+    }
+  }
+
+  private Optional<String> buildConceptIdListClause(
+      Domain domain, List<ConceptSet> conceptSetsSelected) {
+    final Optional<String> conceptSetSqlInClauseMaybe;
+    if (supportsConceptSets(domain)) {
+      conceptSetSqlInClauseMaybe =
+          Optional.of(buildConceptIdSqlInClause(domain, conceptSetsSelected));
+    } else {
+      conceptSetSqlInClauseMaybe = Optional.empty();
+    }
+    return conceptSetSqlInClauseMaybe;
+  }
+
+  private boolean supportsConceptSets(Domain domain) {
+    return domain != Domain.PERSON;
+  }
+
+  private String buildConceptIdSqlInClause(Domain domain, List<ConceptSet> conceptSetsSelected) {
+    final String conceptSetIDs =
+        conceptSetsSelected.stream()
+            .filter(cs -> domain == cs.getDomainEnum())
+            .flatMap(cs -> cs.getConceptIds().stream().map(cid -> Long.toString(cid)))
+            .collect(Collectors.joining(", "));
+    return buildSqlInConstraintList(conceptSetIDs);
+  }
+
+  private String buildSqlInConstraintList(String conceptSetIDs) {
+    if (conceptSetIDs.isEmpty()) {
+      // TODO(jaycarlton): this wont' work if there aren't any concept set IDs b/c SQL IN() list
+      // can't be empty. In the meantime this error means user won't get a SQL syntax error.
+      throw new IllegalStateException("We currently can't handle an empty list of concept IDs");
+    }
+
+    return " IN (" + conceptSetIDs + ")";
+  }
+
+  private QueryJobConfiguration buildQueryJobConfiguration(
+      Map<String, QueryParameterValue> namedCohortParameters, String query) {
+    return bigQueryService.filterBigQueryConfig(
+        QueryJobConfiguration.newBuilder(query)
+            .setNamedParameters(namedCohortParameters)
+            .setUseLegacySql(false)
+            .build());
+  }
 
-      queryMap.put(query, cohortParameters);
-      QueryJobConfiguration queryJobConfiguration =
-          bigQueryService.filterBigQueryConfig(
-              QueryJobConfiguration.newBuilder(query)
-                  .setNamedParameters(cohortParameters)
-                  .setUseLegacySql(false)
-                  .build());
-      dataSetUtil.put(d.toString(), queryJobConfiguration);
+  @VisibleForTesting
+  public boolean conceptSetSelectionIsNonemptyAndEachDomainHasAtLeastOneConcept(
```

Comment:
nit: missing capital "e" in Empty for lowerCamelCase

[Optional] It looks like this method is effectively "isConceptSetValid," but with a very long and explicit method name. If I were bike-shedding this, I'd say maybe the two rules that show up in the method name ("selection is non-empty" and "each domain has nonzero concepts") should each be their own method, which would make for more modular logic and less monolithic method names. But this is so low priority I probably wouldn't bother...